sycl/source/detail/scheduler/scheduler.cpp

```diff
@@ -455,8 +455,15 @@ void Scheduler::cleanupCommands(const std::vector<Command *> &Cmds) {
 
   } else {
     std::lock_guard<std::mutex> Lock{MDeferredCleanupMutex};
-    MDeferredCleanupCommands.insert(MDeferredCleanupCommands.end(),
-                                    Cmds.begin(), Cmds.end());
+    // Full cleanup for fusion placeholder commands is handled by the entry
```

Comment:
Could you elaborate a bit on this? Why are we trying to avoid deferred cleanup while keeping the non-deferred one here? Do we want both the non-deferred cleanup and the mechanism in the fusion entry points to apply?

Reply:
We are fine with the regular cleanup to go ahead. There is an early return [here](https://github.com/intel/llvm/blob/sycl/sycl/source/detail/scheduler/graph_builder.cpp#L1278) in the regular cleanup, so the regular cleanup would not delete the command and can be combined safely with the mechanism in the kernel entry points.

The deferred cleanup on the other hand might lead to access after the lifetime. 
On a high level, the problematic scenario looks as follows:
1. Fusion command gets added to the deferred cleanup list
2. The fusion entry point performs a full cleanup of the command, deleting it.
3. The deferred cleanup happens, goes over the list and tries to cleanup the fusion command that was already deleted. 

By not allowing fusion commands to be added to the deferred cleanup list, we can avoid this scenario. Cleanup still happens through the fusion entry points.

Reply:
> We are fine with the regular cleanup to go ahead. There is an early return [here](https://github.com/intel/llvm/blob/sycl/sycl/source/detail/scheduler/graph_builder.cpp#L1278) in the regular cleanup, so the regular cleanup would not delete the command and can be combined safely with the mechanism in the kernel entry points.

Do we see any benefit to that? It seems that we'll have the post-enqueue cleanup handle fusion commands **sometimes**: they're either fully handled by the fusion-specific cleanup or the work is split between that and the post-enqueue cleanup (where the latter just isolates the node first).

If we can exclude fusion commands from the post-enqueue cleanup altogether (like we do with host tasks now, which have their own cleanup point) without losing anything, I think that would be a cleaner solution.

Reply:
We only need to keep the fusion command around for its event, so we can return that from `complete_fusion` if that is called multiple times. We are not interested in keeping the dependencies of the fusion command around once it gets to post-enqueue cleanup.

Post-enqueue cleanup will also update all dependencies and uses of the fusion command, so I think there is value in having post-enqueue cleanup handle that part of a fusion command.

Reply:
> We only need to keep the fusion command around for its event, so we can return that from complete_fusion if that is called multiple times. We are not interested in keeping the dependencies of the fusion command around once it gets to post-enqueue cleanup.

There's probably something I'm missing here, but if the command is only needed to be kept alive for the event, is it possible to just store the event instead and let the cleanup work with fusion commands as normal?

>Post-enqueue cleanup will also update all dependencies and uses of the fusion command, so I think there is value in having post-enqueue cleanup handle that part of a fusion command.

What happens if it doesn't, i.e. the cleanup is deferred? Are those handled by the fusion-specific cleanup then? If so, is the value in having both that it happens earlier?

Reply:
> > We only need to keep the fusion command around for its event, so we can return that from complete_fusion if that is called multiple times. We are not interested in keeping the dependencies of the fusion command around once it gets to post-enqueue cleanup.
> 
> There's probably something I'm missing here, but if the command is only needed to be kept alive for the event, is it possible to just store the event instead and let the cleanup work with fusion commands as normal?

The fusion command is what gets stored into the [`MFusionMap`](https://github.com/intel/llvm/blob/sycl/sycl/source/detail/scheduler/scheduler.hpp#L753) member of `GraphBuilder` to indicate an ongoing or previously completed fusion. If we want to delete the fusion command normally, we would need to store the element there (e.g., with `std::variant`). I don't think that would be cleaner than storing the rather lightweight `KernelFusionCmd` in there.

> 
> > Post-enqueue cleanup will also update all dependencies and uses of the fusion command, so I think there is value in having post-enqueue cleanup handle that part of a fusion command.
> 
> What happens if it doesn't, i.e. the cleanup is deferred? Are those handled by the fusion-specific cleanup then?

Yes, it is handled, because the fusion-specific cleanup in `cleanUpCmdFusion` eventually also calls `cleanupCommand` [here](https://github.com/intel/llvm/blob/sycl/sycl/source/detail/scheduler/graph_builder.cpp#L1389).

>  If so, is the value in having both that it can happen earlier?

Yes, it can happen earlier and we can also reuse the same logic from the fusion-specific cleanup.